### PR TITLE
Add support for serializing properties in interface hierarchies

### LINKED ISSUE #41749: Serialization of properties from inherited interfaces vs abstract base classes
I would expect that inherited properties of interfaces would be serialized similar to (abstract) base classes.

Given:
 
    public interface IBase
    {
        string BaseProperty { get; set; }
    }

    public interface IDerived : IBase
    {
        string DerivedProperty { get; set; }
    }

    public class Derived : IDerived
    {
        public string BaseProperty { get; set; }
        public string DerivedProperty { get; set; }
    }

    public class Data 
    {
        public IDerived Derived { get; set; }
    }

    Data data = new Data 
    { 
        Derived = new Derived
        { 
            BaseProperty = "B", 
            DerivedProperty = "D" 
        } 
    };
    string json = JsonSerializer.Serialize(data);

Where the json string is missing the BaseProperty:

`{"Derived":{"DerivedProperty":"D"}}`

replacing the interface with an abstract class results in:

`{"Derived":{"DerivedProperty":"D","BaseProperty":"B"}}`

    public abstract class  IBase
    {
        public string BaseProperty { get; set; }
    }

    public abstract class IDerived : IBase
    {
        public string DerivedProperty { get; set; }
    }

    public class Derived : IDerived
    {
    }

    public class Data 
    {
        public IDerived Derived { get; set; }
    }

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #41749.

cc @captainsafia

### CODE REVIEW

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs [hunk 1]
```diff
@@ -324,5 +322,102 @@ public static bool IsKeyValuePair(this Type type, Type? keyValuePairType = null)
 
             return defaultValue;
         }
+
+        /// <summary>
+        /// Returns a list containing the class hierarchy for the given type,
+        /// starting from the current type up to the least derived base type.
+        /// </summary>
+        public static Type[] GetSortedClassHierarchy(this Type type)
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:330]: >Sorted [](http://example.com/codeflow?start=32&length=6)

nit: the word sorted here is a little bit confusing. Maybe something like `GetClassHierarchyChain` be clearer?

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs [hunk 2]
```diff
@@ -324,5 +322,102 @@ public static bool IsKeyValuePair(this Type type, Type? keyValuePairType = null)
...
+        {
+            var results = new List<Type>();
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:332]: >var results = new List<Type>(); [](http://example.com/codeflow?start=12&length=31)

do we care about optimization here? I am asking as we can have a check like 

if (type == typeof(object))
   return Array.Empty<Type>()

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:332]: We could, however `typeof(object)` uses a specialized converter that would never call into this method.

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:330]: The name reflects the fact that the types are topologically sorted. This is not particularly meaningful for classes which always have chained hierarchies like you mention, however the name simply reflects the `GetSortedInterfaceHierarchy` which might not be a chain and returns a topologically sorted sequence.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 3]
```diff
@@ -1026,7 +1026,11 @@ private TypeGenerationSpec GetOrAddTypeGenerationSpec(Type type, JsonSourceGener
 
                         bool propertyOrderSpecified = false;
 
-                        for (Type? currentType = type; currentType != null; currentType = currentType.BaseType)
+                        Type[] sortedTypeHierarchy = type.IsInterface
+                            ? type.GetSortedInterfaceHierarchy()
+                            : type.GetSortedClassHierarchy();
```
Comment (tarekgh) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1031]: would make sense moving these lines into a helper inline method? I am seeing it is repeated somewhere else.

File: src/libraries/System.Text.Json/Common/ReflectionExtensions.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Text.Json/Common/ReflectionExtensions.cs:330]: Yes, I saw that after continue reviewing the code :-)

**Comment by tarekgh** [comment]
```
/_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs(64,13): Trim analysis error IL2072: System.Text.Json.Serialization.Metadata.ReflectionJsonTypeInfo<T>.LateAddProperties(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.Interfaces' in call to 'System.Text.Json.Reflection.ReflectionExtensions.GetSortedInterfaceHierarchy(Type)'. The return value of method 'System.Text.Json.Serialization.Metadata.JsonTypeInfo.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/sfx.proj]

```

**Review by tarekgh — APPROVED** [review]
modulo the analyzer failure, LGTM.

**Comment by eiriktsarpalis** [comment]
> ```
> /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs(64,13): Trim analysis error IL2072: System.Text.Json.Serialization.Metadata.ReflectionJsonTypeInfo<T>.LateAddProperties(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.Interfaces' in call to 'System.Text.Json.Reflection.ReflectionExtensions.GetSortedInterfaceHierarchy(Type)'. The return value of method 'System.Text.Json.Serialization.Metadata.JsonTypeInfo.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/__w/1/s/src/libraries/sfx.proj]
> ```

The changes seem to have triggered false positive warnings being emitted by the trimmer. After a lot of experimentation, I was able to isolate the issue to the one `for` loop that was changed to a `foreach` loop in `ReflectionJsonTypeInfoOfT.cs`. I refactored the `foreach` body into a separate method to avoid hitting the issue.

**Comment by eiriktsarpalis** [comment]
Wasm test failures are unrelated.

**Comment by angelaki** [comment]
Just since I'm curious / fast checking haven't found it: where / how are duplicate property named handled?

**Comment by eiriktsarpalis** [comment]
Same as how they are handled in regular class hierarchies: we throw an exception:

https://github.com/dotnet/runtime/blob/58f59db3f3be886d6c98c9ca6b4ab5fd8b8edda7/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs#L826

**Comment by angelaki** [comment]
Ah sure, Attributes made it possible before, too! Thank you very much.

**Pull request merged**